src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -99,10 +107,117 @@ private static bool TryDecodeSuppressMessageAttributeData (CustomAttribute attri
 
 		private bool IsLocallySuppressed (int id, ICustomAttributeProvider provider, out SuppressMessageInfo info)
 		{
-			Dictionary<int, SuppressMessageInfo> suppressions;
 			info = default;
+			if (provider == null)
+				return false;
+
+			Dictionary<int, SuppressMessageInfo> suppressions;
 			return _localSuppressions.TryGetValue (provider, out suppressions) &&
 				suppressions.TryGetValue (id, out info);
 		}
+
+		private bool IsGloballySuppressed (int id, ICustomAttributeProvider provider, out SuppressMessageInfo info)
+		{
+			DecodeGlobalSuppressMessageAttributes ();
+			return _globalSuppressions.HasGlobalSuppressions (id, provider, out info);
+		}
+
+		private void DecodeGlobalSuppressMessageAttributes ()
+		{
+			if (_globalSuppressions == null) {
+				var suppressions = new GlobalSuppressions ();
+				HashSet<ModuleDefinition> modules = new HashSet<ModuleDefinition> ();
+				foreach (var assembly in _context.Resolver.AssemblyCache.Values) {
```

Comment:
Currently this is "probably" OK since we preload all assemblies really early on. That said if we hit a warning before we finish resolving assemblies this will lead to us possibly missing some global suppressions.

I think a better way to do would be to have a dictionary/hashset of already "initialized" assemblies. And when `IsGloballySuppressed` is called, take the assembly from the `provider` and initialize just that assembly (if we didn't already do that). That way even if we move to lazy-loading assemblies, this will just work. And it's also "lazy" in the sense that we won't go over all assemblies, but only those which generate warnings (which is almost guaranteed to be the set which is marked).

Reply:
Wouldn't this cause a problem when checking for global suppressions? I think we need to go through all assemblies to check if there are any global suppressions specified which could potentially be targeting types or members in other assemblies. I'm thinking in a scenario where an assembly B has the following suppression: 
```C#
[module: UnconditionalSuppressMessage (..., Scope = "type", Target = "A.Foo")]
```
but there's no warning thrown from within B, thus never initializing it's assembly here. Then, if a warning is thrown in the scope of type `A.Foo`, we wouldn't know that this warning was meant to be suppressed.